### PR TITLE
[SPARK-34807][SQL] Move TransposeWindow before Operator push down 

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr move `TransposeWindow` before `Operator push down` to make `PushDownPredicates` push down more cases. For example:
```scala
spark.range(10).selectExpr("id AS a", "id AS b", "id AS c", "id AS d").createTempView("t1")
val df = spark.sql(
  """
    |SELECT *
    |  FROM (
    |    SELECT b,
    |      sum(d) OVER (PARTITION BY a, b),
    |      rank() OVER (PARTITION BY a ORDER BY c)
    |    FROM t1
    |  ) v1
    |WHERE b = 2
    |""".stripMargin).explain(true)
```

Before this pr:
```
== Optimized Logical Plan ==
Project [b#221L, sum(d) OVER (PARTITION BY a, b ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#231L, RANK() OVER (PARTITION BY a ORDER BY c ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)#232]
+- Filter (b#221L = 2)
   +- Window [rank(c#222L) windowspecdefinition(a#220L, c#222L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS RANK() OVER (PARTITION BY a ORDER BY c ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)#232], [a#220L], [c#222L ASC NULLS FIRST]
      +- Project [b#221L, a#220L, c#222L, sum(d) OVER (PARTITION BY a, b ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#231L]
         +- Window [sum(d#223L) windowspecdefinition(a#220L, b#221L, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS sum(d) OVER (PARTITION BY a, b ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#231L], [a#220L, b#221L]
            +- Project [id#218L AS b#221L, id#218L AS d#223L, id#218L AS a#220L, id#218L AS c#222L]
               +- Range (0, 10, step=1, splits=Some(2))
```
After this pr:
```
== Optimized Logical Plan ==
Project [b#221L, sum(d) OVER (PARTITION BY a, b ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#231L, RANK() OVER (PARTITION BY a ORDER BY c ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)#232]
+- Window [sum(d#223L) windowspecdefinition(a#220L, b#221L, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS sum(d) OVER (PARTITION BY a, b ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#231L], [a#220L, b#221L]
   +- Project [b#221L, d#223L, a#220L, RANK() OVER (PARTITION BY a ORDER BY c ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)#232]
      +- Filter (b#221L = 2)
         +- Window [rank(c#222L) windowspecdefinition(a#220L, c#222L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS RANK() OVER (PARTITION BY a ORDER BY c ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)#232], [a#220L], [c#222L ASC NULLS FIRST]
            +- Project [id#218L AS b#221L, id#218L AS d#223L, id#218L AS a#220L, id#218L AS c#222L]
               +- Range (0, 10, step=1, splits=Some(2))
```

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q47.sf100/explain.txt [hunk 1]
```diff
@@ -1,62 +1,61 @@
...
++- * Project (56)
+   +- * SortMergeJoin Inner (55)
+      :- * Project (47)
+      :  +- * SortMergeJoin Inner (46)
+      :     :- * Sort (36)
+      :     :  +- Exchange (35)
+      :     :     +- * Project (34)
+      :     :        +- * Filter (33)
+      :     :           +- Window (32)
+      :     :              +- * Filter (31)
```
Comment (wangyum) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q47.sf100/explain.txt:12]: image

**Comment by tanelk** [comment]
Reordering the optimizers seems like a bit of a band-aid fix. 
If I'm not mistaken the issue is similar to the one I tried to fix in #31677. The `CollapseWindow` rule did not work with `Project` nodes between windows. It seems, that the `TransposeWindow` has a similar problem.

@cloud-fan did comment (https://github.com/apache/spark/pull/31677#issuecomment-800894675), that perhaps we should merge `Project` and `Window` nodes. But that seems like a major refacto of everything related to the logical `Window` nodes.

**Comment by tanelk** [comment]
@wangyum, I'll answer your question (https://github.com/apache/spark/pull/31677#pullrequestreview-616975279) here.
Just to clarify - the #31677 does not fix the issue, you are trying to fix here. But they are very similar. I improved the `CollapseWindow`,  the `TransposeWindow` can be improved in a similar way to fix your issue.

I tried a quick change:
```diff
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
index 3e3550d5da..e629ccc268 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
@@ -991,13 +991,24 @@ object TransposeWindow extends Rule[LogicalPlan] {
     })
   }

+  private def windowsCompatible(w1: Window, w2: Window): Boolean = {
+    w1.references.intersect(w2.windowOutputSet).isEmpty &&
+      w1.expressions.forall(_.deterministic) &&
+      w2.expressions.forall(_.deterministic) &&
+      compatiblePartitions(w1.partitionSpec, w2.partitionSpec)
+  }
+
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
-    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
-        if w1.references.intersect(w2.windowOutputSet).isEmpty &&
-           w1.expressions.forall(_.deterministic) &&
-           w2.expressions.forall(_.deterministic) &&
-           compatiblePartitions(ps1, ps2) =>
-      Project(w1.output, Window(we2, ps2, os2, Window(we1, ps1, os1, grandChild)))
+    case w1 @ Window(_, _, _, w2 @ Window(_, _, _, grandChild))
+        if windowsCompatible(w1, w2) =>
+      Project(w1.output, w2.copy(child = w1.copy(child = grandChild)))
+
+    case w1 @ Window(_, _, _, Project(pl, w2 @ Window(_, _, _, grandChild)))
+      if windowsCompatible(w1, w2) && w1.references.subsetOf(grandChild.outputSet) =>
+      Project(
+        pl ++ w1.windowOutputSet,
+        w2.copy(child = w1.copy(child = grandChild))
+      )
   }
 }
``` 

And it changes the TPC-DS q47 and TPC-DS q57 in the same way your PR does, but I find this change to be more robust.

**Comment by wangyum** [comment]
@tanelk Could you create pr for your changes: https://github.com/apache/spark/pull/31907#issuecomment-803521361. I'm close this pr.

**Pull request closed without merging**